src/main/scala/viper/silver/ast/utility/chopper/Chopper.scala

```diff
@@ -0,0 +1,1137 @@
+
+    val fastEdges = Array.fill(N)(mutable.SortedSet.empty[Int])
+    for ((l, r) <- edges) {
+      fastEdges(l).add(r)
+    }
+
+    val setOfVerticesToProgram = inverse(program)
+    val setOfIdsToProgram = (set: ViperGraphs.Program) => setOfVerticesToProgram(set map idToVertex)
+
+    new ViperGraph(
+      numberOfNodes = N,
+      importantNodes = importantNodes,
+      edges = fastEdges,
+      toVertex = idToVertex,
+      toVpr = setOfIdsToProgram,
+    )
+  }
+}
+
+
+
+
+object Vertices {
+
+  /**
+    * Abstract the smallest parts of a Viper program and provides necessary information to compute merge penalties.
+    * There are Vertex instances for each Viper member, except domains, which are split into domain types, functions, and axioms.
+    * Furthermore, [[Vertices.MethodSpec]] and [[Vertices.PredicateSig]] represent methods and predicates without body, respectively.
+    * */
+  sealed trait Vertex
+
+  /** Represents a Viper method member. */
+  case class Method private[Vertices](methodName: String) extends Vertex
+
+  /** Represents a Viper method member without the body. */
+  case class MethodSpec(methodName: String) extends Vertex
+
+  /** Represents a Viper function member. */
+  case class Function(functionName: String) extends Vertex
+
+  /** Represents a Viper predicate member. */
+  case class PredicateSig(predicateName: String) extends Vertex
+
+  /** Represents a Viper predicate member without the body. */
+  case class PredicateBody private[Vertices](predicateName: String) extends Vertex
+
+  /** Represents a Viper field member. */
+  case class Field(fieldName: String) extends Vertex
+
+  /** Represents a Viper domain function declaration. */
+  case class DomainFunction(funcName: String) extends Vertex
+
+  /** Represents a Viper domain axiom declaration. */
+  case class DomainAxiom(v: ast.DomainAxiom, d: ast.Domain) extends Vertex
+
+  /** Represents a Viper domain type declaration. */
+  case class DomainType(v: ast.DomainType) extends Vertex
+
+  /**
+    * Represents a dependency that must be included in all programs.
+    * If something always has to be included, then it is a dependency of Always
+    * */
+  case object Always extends Vertex
+
+  // the creation of the following vertices has special cases and should not happen outside of the Vertex object
+  object Method {
+    private[Vertices] def apply(methodName: String): Method = new Method(methodName)
+  }
+
+  object PredicateBody {
+    private[Vertices] def apply(predicateName: String): PredicateBody = new PredicateBody(predicateName)
+  }
+
+  /** This function is only allowed to be called in the following cases:
+    * 1) applying [[Vertices.toDefVertex]] to the predicate referenced by `predicateName` returns a [[Vertices.PredicateBody]] instance.
+    * 2) The result is used as the target of a dependency.
+    * */
+  def predicateBody_check_that_call_is_ok(predicateName: String): PredicateBody = Vertices.PredicateBody(predicateName)
```

Comment:
I think this name is misleading, as it suggests that it performs a check. Besides, it looks very weird in Scala (where people tend to go with camel case). I would rename this `unsafeGetPredicateBody`